### ISSUE TITLE
Tertiary Provider extensions

### ISSUE THREAD

**Issue by smudge202** [smudge202/compose#83]
We add providers for interfaces typically through Service Extensions created by the provider. These service extensions will take form in one of two ways:

#### `Add`

If the service extension simply adds itself to the `IServiceCollection`, the MS provider (at time of writing) will resolve duplicated bindings using the **last** added binding.

i.e.

```c#
services
  .AddProvider1()
  .AddProvider2(); // provider 2 will be used
```

#### `TryAdd`

If the service extension uses the `TryAdd`  extension, it will only add a service descriptor if one does not already exist, meaning the **first** binding added will be used.

i.e.

```c#
services
  .AddProvider1() /* provider 1 will be used */
  .AddProvider2();
```

As you can hopefully see from the example, there is no way of knowing from the application host which of the providers will be used.

When performing a _Transition_, Compose will attempt to retrieve instances of the new implementation using the underlying provider, allowing the new implementations to utilise dependency injection. However, given the above extension behaviour, it is difficult for consumers to add both providers into the Service Collection during composition, because of the ambiguity in provider selection.

I propose providing an API that allows the end user to have multiple service collections (potentially one per provider). There is of course the consideration of shared components between the service collections though. I'd be interested to hear some thoughts on this, and particularly any suggestions regarding an API that can achieve the above?

**Comment by herecydev** [comment]
![image](https://cloud.githubusercontent.com/assets/11328618/8110830/386fe688-1056-11e5-9328-7760375ac458.png)

:shipit:

**Comment by smudge202** [comment]
Discussed this with @herecydev and came up with an idea. The idea is to provide clean service collections for use when adding additional providers. Given we're working with an `IAnimal`, and we have 3 providers, `Dog`, `Cat`, and `Fish`, and that we want to use Dog by default and transition to the others

```c#
app.UseServices(services =>
{
  services.AddDog();
  services.AddLogging(config => config.WriteTo.ColouredConsole());
});

app.AddProvider<IAnimal>(services =>
{
  services.AddCat();
  // the logging added above will not be used by Cat, so if you need logging
  // you need to add it here again, but that does allow you to add different
  // logging by provider?
});

// when we want to transition:
app.Transition<IAnimal, Cat>();
// as per herecydevs screen clip, this is easily achievable using the 
// TransitionServices property he suggested

//  I'm also interested in the following construct, which may be better?
var fishProvider == app.AddProvider<IAnimal>(services => services.AddFish());

app.Transition(fishProvider);
```

Thoughts? @herecydev particularly to that last construct. The thoughts behind it are that the service extension may have `Fish` as an internal class, which therefore wouldn't work with the existing API. Doing it as per my last suggestion allows services not to expose their implementations, though doesn't work with the construct you suggested.

**Comment by herecydev** [comment]
I like it! I think how/where the service collections are stored may change through development.

**Comment by sblackler** [comment]
I'm not quiet sure I agree with the comment in `AddProvider<IAnimal>`. If I don't specify a dependency, I would still expect it to resolve to the current one. Overriding the dependency on transition is fine, but if none is supplied then use the original? So given Dog as the base:

- If I specify Cat with no logging, I should get the logger registered with dog
- If I specify Cat with logging, I should get the logger I asked for

**Comment by smudge202** [comment]
That's a lovely idea, @sblackler . One I considered. But to achieve that, we would need to implement our own `Microsoft.Framework.DependencyInjection`. The value gained versus effort required is in poor balance at this stage.

Hmmm... might smoke on it, but there **may** be a way to do what you want. Will outline what I have in mind shortly and you guys can try to break the concept.

**Comment by sblackler** [comment]
Why would you need to do that? I think the behaviour as explained above will cause confusion as people would expect the original dependency unless transitioned. 

On the naming side, wouldn't `AddTransitionGroup<T>` be more appropriate as you are defining more than just the provider `T`?

**Comment by herecydev** [comment]
I think that assumption is dangerous. I specify a logger for "these services". You don't know how that could impact different services. If you're switching to a different set of services, you might have different sinks for logging to also.

**Comment by smudge202** [comment]
Interesting that we're already deriving different expectations. We certainly need to working on clarity regarding naming / API.

As for your question @sblackler , the only way to resolve the behavioral ambiguity described in the OP, is to have the providers be added to different `IServiceCollection`s. I don't think it's a good idea to then merge those collections, because the two providers might interfere with one another's bindings.

When descriptors exist in seperate `IServiceCollection`s, no single `IServiceProvider` is able to service requests using both collections. They can only look at one; thus we'd need to roll our own in order to achieve that.

**Comment by smudge202** [comment]
Regarding merging collections... that **might** be a way to make it work. Need to think on it some more.

**Comment by herecydev** [comment]
What about `.PreserveServices()` to indicate that *where possible* existing services should be resolved to the transitioned services, i.e. reusing the `ILogger`.

This gives control back to the developer. If they don't specify this, they are then forced to provide the additional services. In the case of a data center switch, this is advantageous as they can switch out their `ILogger` from azure to aws for example when they also transition `Dog` to `Cat`.

**Comment by smudge202** [comment]
Wondering if we should go back to having everything inside the `UseServices` block:

```c#
app.UseServices(services =>
{
  services.AddDog();
  services.AddSharedServices(shared => shared.AddLogging());
  services.AddProvider(provider => provider.AddCat());
});
```

**Comment by herecydev** [comment]
Then `IAnimal` has a single implementation only within the `IServiceCollection` surely?

**Comment by smudge202** [comment]
My mistake, the above doesn't use the improved pattern for Adding providers...

Basically, we need an API that can define a default provider, shared components, and tertiary providers.

@herecydev the first example above has two service collections, one called `services`, the other called `provider`. (Again, a clarity issue, but that API doesn't work anyway).

**Comment by herecydev** [comment]
I prefer

```c#
app.UseServices(services =>
{
  services.AddDog();
  services.AddLogging();
});

app.TranstitionServices(backupServices =>
{
  backupServices.AddCat();
}).ShareServices();
// logger shared

 ---- or -----

app.TranstitionServices(backupServices =>
{
  backupServices.AddCat();
  // aws logging
  backupServices.AddLogging();
});
```

**Comment by smudge202** [comment]
Remove the superfluous `ShareServices` method and you derive the expectation Stu @sblackler .

**Comment by smudge202** [comment]
Discussed this over lunch with @sblackler and @herecydev , whom can hopefully confirm the following summary as the outcome and viable solution:

* `app.UseServices` is used to define default services, and is considered the _primary_ collection hereafter.
* Tertiary providers can be added through `AddProvider`, `AddTransition`, or a similar extension (method name TBC). These will create a _tertiary_ service collection, which we will merge the _primary_ collection into using `TryAdd`

This means services in the primary collection are shared by default, but can be overriden by defining the services when adding the tertiary provider (`TryAdd` will not add a service if the `ServiceType` already exists in the collection).

### Issues

* Singletons would not be persisted across the collections/providers. 

Therefore, when merging the primary collection into a tertiary collection, we will modify any `ServiceDescriptor` that uses `ImplementationType` and has a `LifeSpan` of `Singleton` to a new Descriptor for the same `ServiceType`, but using the `ImplementationFactory` and lambda closure to encapsulate `primaryProvider.GetService<TheSingleton>()`.

* Greedy constructor collisions

Most Service Providers (though the MS provider doesn't last time I checked) determine the constructor to use based on a _greedy_ algorithm. That is, they will find the constructor with the most arguments that they are able to fulfill, and pick that one. If a tertiary provider binds a class that has multiple constructors, the shared services might mean a larger constructor is utilised. I think this should be _expected_ behaviour, though will need some documentation because it may not be entirely obvious given the API? I'm open to other suggestions here, but I don't think this is something we _should_ try to resolve, just something to be aware of?

**Comment by sblackler** [comment]
Agreed I dont think that it is something that we should try to resolve. However, the first issue/solution appears fine. Will need a good amount of testing so that we dont break anything though :)

**Comment by smudge202** [comment]
Resolved by d8694a4922cd9f5c5f1bb1e0816632c8f6aa89d9

**Comment by smudge202** [comment]
Oops, this issue isn't resolved yet!

**Comment by smudge202** [comment]
C'mon then, I'm in the mood to code. Let's get this bad boy rolling... (inb4 drunken commit breaks appveyor - the site, not the build).
